### PR TITLE
Fix/pandas client performance

### PR DESCRIPTION
**Pull request by sperretcraft**
Some (very) little changes to make things a little faster !

As I removed the `apply`, I removed dummy columns needed to avoid an unwanted behavior.

According to the benckmark I did, these little improvements make things (almost) 10 times faster :+1: 
benchmark

P.S. To make further optimization, I need to discuss about the missing values, the optional values and the timezone-related features :-)

### CODE REVIEW

File: craftai/pandas/interpreter.py [hunk 1]
```diff
@@ -2,64 +2,45 @@
...
+    }
   except CraftAiNullDecisionError as e:
-    return pd.Series(data=[e.message], index=["error"])
+    return {"error": e.message}
 
 class Interpreter(VanillaInterpreter):
   @staticmethod
   def decide_from_contexts_df(tree, contexts_df):
     _, configuration, _ = VanillaInterpreter._parse_tree(tree)
+    df = contexts_df.copy(deep=True)
```
Comment (Totoketchup) [inline craftai/pandas/interpreter.py]: Why do you need a copy here ?

Comment (sperretcraft) [reply craftai/pandas/interpreter.py]: This method may add a timezone column to `df`. The copy is needed not to modify the dataframe parameter of the method in the method itself.

Comment (Totoketchup) [reply craftai/pandas/interpreter.py]: what about using df = df.assign([tzcol] = ... )

Comment (sperretcraft) [reply craftai/pandas/interpreter.py]: After checking, `pandas.DataFrame.assign` source code contains a `df.copy(deep=True)` statement. So I guess it doesn't make a big difference.

Comment (Totoketchup) [reply craftai/pandas/interpreter.py]: Yep, then better to leave it like you did ;)

File: tests/test_pandas.py [hunk 2]
```diff
@@ -152,6 +155,30 @@ def test_decide_from_contexts_df():
...
+  df = CLIENT.decide_from_contexts_df(tree, test_df)
+
+  assert_true(isinstance(df, pd.DataFrame))
+  assert_equal(len(df), 0)
+  assert_true(test_df.equals(test_df_copy))
+
+@with_setup(setup_complex_agent_with_data, teardown)
+def test_decide_from_contexts_df_duplicated_index():
+  tree = CLIENT.get_decision_tree(AGENT_ID, COMPLEX_AGENT_DATA.last_valid_index().value // 10 ** 9)
+  #test_df = COMPLEX_AGENT_DATA.iloc[[1, 1, 1, 2], :] # make the test fail
```
Comment (Totoketchup) [inline tests/test_pandas.py]: Why this one is failing and not the next one ?

Comment (sperretcraft) [reply tests/test_pandas.py]: I'm not sure, but it's related with the timezone thing i guess.

Comment (sperretcraft) [reply tests/test_pandas.py]: I would like to deal with this aspect in a other PR.

Comment (Totoketchup) [reply tests/test_pandas.py]: Then better to add this test in another PR I guess

Comment (sperretcraft) [reply tests/test_pandas.py]: ok

File: craftai/pandas/interpreter.py [hunk 3]
```diff
@@ -2,64 +2,52 @@
...
-    keys, values = zip(*[
-      (output + "_" + key, value)
-      for output, output_decision in decision["output"].items()
-      for key, value in output_decision.items()
-    ])
-
-    return pd.Series(data=values, index=keys)
-  except CraftAiNullDecisionError as e:
-    return pd.Series(data=[e.message], index=["error"])
+from .utils import is_valid_property_value, create_timezone_df, format_input
```
Comment (sperretcraft) [inline craftai/pandas/interpreter.py]:
```suggestion
from .utils import create_timezone_df, format_input
```

File: craftai/interpreter.py [hunk 4]
```diff
@@ -12,34 +12,43 @@ class Interpreter(object):
...
+    decision = interpreter.decide(configuration, bare_tree, context)
+    decision["context"] = context
+
+    return decision
+
+  @staticmethod
+  def _get_interpreter(tree_version):
     if semver.match(tree_version, ">=1.0.0") and semver.match(tree_version, "<2.0.0"):
-      decision = InterpreterV1.decide(configuration, bare_tree, context)
+      interpreter = InterpreterV1
```
Comment (Totoketchup) [inline craftai/interpreter.py]:
```suggestion
      return InterpreterV1
```
I guess you can directly return it here

File: craftai/interpreter.py [hunk 5]
```diff
@@ -12,34 +12,43 @@ class Interpreter(object):
...
     elif semver.match(tree_version, ">=2.0.0") and semver.match(tree_version, "<3.0.0"):
-      decision = InterpreterV2.decide(configuration, bare_tree, context)
+      interpreter = InterpreterV2
```
Comment (Totoketchup) [inline craftai/interpreter.py]:
```suggestion
      return InterpreterV2
```
Same here

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,16 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased](https://github.com/craft-ai/craft-ai-client-python/compare/v1.15.2...HEAD) ##
+##
+
+### Added
+
+- Some tests to pandas client.
```
Comment (Totoketchup) [inline CHANGELOG.md]:
```suggestion
```
Not necesseray to add this. Manly talk about features addition.

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,16 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+- Some subfunctions in `craftai.interpreter`.
```
Comment (Totoketchup) [inline CHANGELOG.md]:
```suggestion
```
Same here.

File: craftai/interpreter.py [hunk 8]
```diff
@@ -12,34 +12,43 @@ class Interpreter(object):
...
       )
-
-    decision["context"] = context
-
-    return decision
-
-  ####################
-  # Internal helpers #
-  ####################
+    return interpreter
```
Comment (Totoketchup) [inline craftai/interpreter.py]:
```suggestion
```
you don't need this one anymore with the returns

File: tests/test_pandas.py [hunk 9]
```diff
@@ -152,6 +155,17 @@ def test_decide_from_contexts_df():
 
   assert_equal(output["output"]["b"]["predicted_value"], "Pierre")
 
+@with_setup(setup_complex_agent_with_data, teardown)
+def test_decide_from_contexts_df_zero_rows():
+  tree = CLIENT.get_decision_tree(AGENT_ID, COMPLEX_AGENT_DATA.last_valid_index().value // 10 ** 9)
+  test_df = COMPLEX_AGENT_DATA.iloc[:0, :]
```
Comment (Totoketchup) [inline tests/test_pandas.py]:
```suggestion
  test_df = pd.Dataframe([])
```
What about just using something like this ?

Comment (sperretcraft) [reply tests/test_pandas.py]: It's doesn't test the same thing ! The current test_df has some columns whereas the suggested one doesn't.

Both could be interesting to test but I am not sure about the expected behavior of the function on the suggested dataFrame.

Comment (Totoketchup) [reply tests/test_pandas.py]: Ah right ! Let's leave in like this then

**Pull request merged**